doc/notes/188_enhance_logging_ant.adoc

```diff
@@ -0,0 +1,245 @@
+= Enhance Logging with `log4cplus`
+
+xtUML Project Analysis Note
+
+== 1 Abstract
+
+This note outlines enhancements to the practice of logging in the Munin
+Protocol Verifier application.
+
+== 2 Introduction and Background
+
+Until January 2024, the MASL C{plus}{plus} model compiler <<dr-2>> used POCO
+<<dr-3>> as the underlying logging technology.  Now MASL uses
+`log4cplus` <<dr-4>>.  `log4cplus` provides more functionality and is
+consistent with `log4j`, which is a popular logging framework for Java.
+
+During testing of the Protocol Verifier, it was noted that the POCO logger
+was using more memory than expected.  This was the trigger to change
+logging technologies.
+
+`log4cplus` is a bit different.  It makes sense to revisit the logging
+strategy employed by the Munin project in the Protocol Verifier.  This
+note will outline a Way Forward.
+
+The primary requirements for what to log come from the requirements for
+reporting which are documented in <<dr-5>>.  It is noted that logging and
+reporting are closely related.  If requirements change for reporting, it
+is helpful if a configuration change to the logging can supply the needed
+extra reporting information.
+
+=== 2.1 Logging Best Practices
+
+As preparation for enhancing logging, a few published "Best Practice"
+guides were studied.  We will take the opportunity in Munin to adopt/apply
+some of these practices while we are changing to the `log4*` idiom of
+logging.
+
+The first Best Practice to logging is to author a mission statement for
+the logging and proceed from there.  This aligns with Shlaer-Mellor
+domain modelling quite nicely.
+
+== 3 Requirements
+
+. Establish a Project Munin logging mission statement to guide decision
+  making.
+. Make reporting flow from logging.
+. Support logging levels:
+  ** Fatal
+  ** Error
+  ** Warning
+  ** Information
+  ** Debug
+  ** Trace
+. Support conditional log message string arithmetic.
+. Establish conventions for enabled logging.
+. Support logger hierarchy.
+. Establish naming conventions for logger hierarchy.
+. Establish rules and guidelines for the construction of log messages.
+. Facilitate machine parsing of log messages (using structured logging).
+  TODO: Consider the format that Prometheus supports.
+. Identify tooling that will be used to harvest the log data.
+. Protect privacy and security by logging application-level information
+  judiciously.
+
+== 4 Analysis
+
+=== 4.1 Munin Logging Mission Statement
+
+* Munin software shall log information critical to the verification of protocols.
+  This includes detection of alarms, failures and successes.
+* Munin software shall be instrumented to enable the measurement of the
+  performance of the Protocol Verifier application itself (benchmarking).
+* Munin software shall be instrumented sparingly with debug statements
+  adequate for defect isolation (debugging).
```

Comment:
I conclude from this that we will, in some cases, remove existing debug messages - because it is considered more than what is needed to satisfy the 'adequate' requirement - is that the case?